### ISSUE TITLE
Child elements overlapping

### ISSUE THREAD

**Issue by douglasjunior** [ApmeM/android-flowlayout#39]
I'm trying to use the "flow layout" to group "TextView" but I'm having overlapping problem.

See in the image:
![captura de tela 2015-04-08 as 08 57 59](https://cloud.githubusercontent.com/assets/1512341/7044863/4ad04e18-ddce-11e4-8ff9-c064f9a3e2b1.png)

My layout:
```xml
<?xml version="1.0" encoding="utf-8"?>
<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:f="http://schemas.android.com/apk/res-auto"
    android:orientation="vertical"
    android:layout_width="fill_parent"
    android:layout_height="fill_parent">

    <LinearLayout
        android:orientation="horizontal"
        android:layout_width="fill_parent"
        android:layout_height="wrap_content"
        android:layout_marginTop="8dp"
        android:layout_marginBottom="8dp"
        android:gravity="center_vertical">

        <LinearLayout
            android:orientation="vertical"
            android:layout_width="wrap_content"
            android:layout_height="wrap_content"
            android:layout_gravity="center_vertical">

            <TextView
                android:layout_width="wrap_content"
                android:layout_height="wrap_content"
                android:textAppearance="?android:attr/textAppearanceLarge"
                android:text="13:50"
                android:id="@+id/textView6"
                android:textSize="35sp"
                android:gravity="center" />

            <TextView
                android:layout_width="wrap_content"
                android:layout_height="wrap_content"
                android:textAppearance="?android:attr/textAppearanceMedium"
                android:text="30 min"
                android:id="@+id/textView5"
                android:layout_gravity="center_horizontal" />
        </LinearLayout>

        <LinearLayout
            android:orientation="vertical"
            android:layout_width="fill_parent"
            android:layout_height="fill_parent"
            android:layout_weight="1"
            android:gravity="center">

            <LinearLayout
                android:orientation="vertical"
                android:layout_width="wrap_content"
                android:layout_height="wrap_content">

                <LinearLayout
                    android:orientation="horizontal"
                    android:layout_width="wrap_content"
                    android:layout_height="wrap_content">

                    <TextView
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content"
                        android:textAppearance="?android:attr/textAppearanceMedium"
                        android:text="Prog. 01"
                        android:id="@+id/tvProgramacao"
                        android:layout_weight="1" />
                </LinearLayout>

                <org.apmem.tools.layouts.FlowLayout
                    f:debugDraw="true"
                    f:layoutDirection="ltr"
                    android:gravity="left"
                    android:layout_width="wrap_content"
                    android:layout_height="wrap_content">

                    <LinearLayout
                        android:layout_margin="5dp"
                        android:orientation="horizontal"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content">

                        <TextView
                            android:layout_width="wrap_content"
                            android:layout_height="wrap_content"
                            android:textAppearance="?android:attr/textAppearanceMedium"
                            android:text="D"
                            android:singleLine="true"
                            android:id="@+id/tvDomingo" />
                    </LinearLayout>

                    <LinearLayout
                        android:layout_margin="5dp"
                        android:orientation="horizontal"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content">

                        <TextView
                            android:layout_width="wrap_content"
                            android:layout_height="wrap_content"
                            android:textAppearance="?android:attr/textAppearanceMedium"
                            android:text="S"
                            android:singleLine="true"
                            android:id="@+id/tvSegunda" />
                    </LinearLayout>

                    <LinearLayout
                        android:layout_margin="5dp"
                        android:orientation="horizontal"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content">

                        <TextView
                            android:layout_width="wrap_content"
                            android:layout_height="wrap_content"
                            android:textAppearance="?android:attr/textAppearanceMedium"
                            android:text="T"
                            android:singleLine="true"
                            android:id="@+id/tvTerca" />
                    </LinearLayout>

                    <LinearLayout
                        android:layout_margin="5dp"
                        android:orientation="horizontal"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content">

                        <TextView
                            android:layout_width="wrap_content"
                            android:layout_height="wrap_content"
                            android:textAppearance="?android:attr/textAppearanceMedium"
                            android:text="Q"
                            android:singleLine="true"
                            android:id="@+id/tvQuarta" />
                    </LinearLayout>

                    <LinearLayout
                        android:layout_margin="5dp"
                        android:orientation="horizontal"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content">

                        <TextView
                            android:layout_width="wrap_content"
                            android:layout_height="wrap_content"
                            android:textAppearance="?android:attr/textAppearanceMedium"
                            android:text="Q"
                            android:singleLine="true"
                            android:id="@+id/tvQuinta"  />
                    </LinearLayout>

                    <LinearLayout
                        android:layout_margin="5dp"
                        android:orientation="horizontal"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content">

                        <TextView
                            android:layout_width="wrap_content"
                            android:layout_height="wrap_content"
                            android:textAppearance="?android:attr/textAppearanceMedium"
                            android:text="S"
                            android:singleLine="true"
                            android:id="@+id/tvSexta" />
                    </LinearLayout>

                    <LinearLayout
                        android:layout_margin="5dp"
                        android:orientation="horizontal"
                        android:layout_width="wrap_content"
                        android:layout_height="wrap_content">

                        <TextView
                            android:layout_width="wrap_content"
                            android:layout_height="wrap_content"
                            android:textAppearance="?android:attr/textAppearanceMedium"
                            android:text="S"
                            android:singleLine="true"
                            android:id="@+id/tvSabado" />
                    </LinearLayout>
                </org.apmem.tools.layouts.FlowLayout>
            </LinearLayout>
        </LinearLayout>

        <LinearLayout
            android:orientation="horizontal"
            android:layout_width="wrap_content"
            android:layout_height="fill_parent"
            android:layout_gravity="center"
            android:padding="10dp">

            <LinearLayout
                android:orientation="horizontal"
                android:layout_width="1dp"
                android:layout_height="fill_parent"
                android:background="#ffffffff"
                android:layout_weight="1"></LinearLayout>
        </LinearLayout>

        <LinearLayout
            android:orientation="vertical"
            android:layout_width="wrap_content"
            android:layout_height="wrap_content">

            <ImageView
                android:layout_width="wrap_content"
                android:layout_height="wrap_content"
                android:id="@+id/imageView"
                android:src="@drawable/ic_filter_white_48dp"
                android:layout_gravity="center_vertical"
                android:layout_marginTop="10dp"
                android:layout_marginRight="10dp"
                android:layout_marginBottom="10dp"
                android:layout_marginLeft="5dp" />

        </LinearLayout>
    </LinearLayout>
</LinearLayout>
```
**I'm forgetting some property?**

**Comment by ApmeM** [comment]
HI, sorry, I cant find where are they overlapping.
As I can see from code and image flowlayout is using only for letters "DSTQQS", the other image parts are using different linear layouts.
As I can see those letters are not overlapping each other, and just did not fit in parent container (or this is a problem?) - maybe because of its size.
Can you provide more details (and remove information that is not related to the problem).

**Comment by douglasjunior** [comment]
I need to display 7 letters within the FlowLayout (D S T Q Q S S), and if you look at the screenshot you will see that the last letter is overlapped, it should be move to the second line, should not it?

Apparently the problem is with the margin. I put the letters in a "LinearLayout" to test, because only the "TextField" had the same problem.

I think I got a way to work. I kept the LinearLayout and put the margin in TextField.

```xml
<LinearLayout
  android:orientation="horizontal"
  android:layout_width="wrap_content"
  android:layout_height="wrap_content">
    <TextView
      android:layout_width="wrap_content"
      android:layout_height="wrap_content"
      android:textAppearance="?android:attr/textAppearanceMedium"
      android:text="Dom."
      android:singleLine="true"
      android:id="@+id/tvDomingo"
      android:layout_margin="5dp" />
</LinearLayout>
```
Looking at the source code of the FlowLayout, I noticed that he measures the child's size to see if it should be added to the current row or the next.

Apparently, when the margin was in LinearLayout, this measures was not correct.

**Comment by ApmeM** [comment]
Hi, I have found, that it is reproducable for this layout as well:

    <?xml version="1.0" encoding="utf-8"?>
    <org.apmem.tools.layouts.FlowLayout
        xmlns:android="http://schemas.android.com/apk/res/android"
        xmlns:f="http://schemas.android.com/apk/res-auto"
        f:debugDraw="true"
        android:layout_width="wrap_content"
        android:layout_height="wrap_content">

        <TextView
            android:layout_margin="5dp"
            android:layout_width="wrap_content"
            android:layout_height="wrap_content"
            android:text="D"/>

    </org.apmem.tools.layouts.FlowLayout>

and the problem was in spacings calculations.
I have fixed it and also create tests.

**Comment by douglasjunior** [comment]
@ApmeM Thank you for your dedication.

Off: I'm curious. You have planned to convert the FlowLayout as a LayoutManager for RecyclerView?

**Comment by dadino** [comment]
When will you release the updated library?

**Comment by ApmeM** [comment]
@douglasjunior : before your comment I even did not know about RecyclerView and it LayoutManagers. Therefore I did not have such plans, but now I will check it. I have created another issue: Support for 'RecyclerView' #59. 
Can you please add example of usage or any code that is already implementing this feature to #59 issue so I can check and apply it to this FlowLayout?

@dadino : I was going to fix #50 issue before release. I think release will be today or tomorrow.
